File: Tests/SignalR.Extras.Autofac.Test/LifetimeHubFacts.cs
using System;
using System.Linq;

using SignalR.Extras.Autofac.Test.Stubs;

using Xunit;

using static SignalR.Extras.Autofac.Test.TestHelper;


namespace SignalR.Extras.Autofac.Test
{

	public class LifetimeHubFacts
	{

		[Fact]
		public void SignalRResolvesValidLifetimeHubs()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			using (var hub = hubManager.ResolveHub(nameof(DynamicLifetimeHubStub))) {
				Assert.IsType<DynamicLifetimeHubStub>(hub);
				Assert.True(hub is ILifetimeHub);
				Assert.NotNull(hub);
			}

			using (var hub = hubManager.ResolveHub(nameof(GenericLifetimeHubStub))) {
				Assert.IsType<GenericLifetimeHubStub>(hub);
				Assert.True(hub is ILifetimeHub);
				Assert.NotNull(hub);
			}
		}


		[Fact]
		public void SignalRResolvesValidOrdinaryHubs()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			using (var hub = hubManager.ResolveHub(nameof(DynamicOrdinaryHubStub))) {
				Assert.IsType<DynamicOrdinaryHubStub>(hub);
				Assert.False(hub is ILifetimeHub);
				Assert.NotNull(hub);
			}

			using (var hub = hubManager.ResolveHub(nameof(GenericOrdinaryHubStub))) {
				Assert.IsType<GenericOrdinaryHubStub>(hub);
				Assert.False(hub is ILifetimeHub);
				Assert.NotNull(hub);
			}
		}


		[Fact]
		public void SignalRResolvesCorrectNumberOfRegisteredHubs()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			var hubs = hubManager.ResolveHubs();
			Assert.Equal(4, hubs.Count());
		}


		[Fact]
		public void SignalRResolvesNewHubInstanceEachTime()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			using (var hub1 = (ILifetimeHub)hubManager.ResolveHub(nameof(DynamicLifetimeHubStub)))
			using (var hub2 = (ILifetimeHub)hubManager.ResolveHub(nameof(DynamicLifetimeHubStub))) {
				Assert.NotSame(hub1, hub2);
			}

			using (var hub1 = (ILifetimeHub)hubManager.ResolveHub(nameof(GenericLifetimeHubStub)))
			using (var hub2 = (ILifetimeHub)hubManager.ResolveHub(nameof(GenericLifetimeHubStub))) {
				Assert.NotSame(hub1, hub2);
			}
		}


		[Fact]
		public void ReceivesNewInstancesOfScopedDependencies()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			using (var hub1 = (DynamicLifetimeHubStub)hubManager.ResolveHub(nameof(DynamicLifetimeHubStub)))
			using (var hub2 = (DynamicLifetimeHubStub)hubManager.ResolveHub(nameof(DynamicLifetimeHubStub))) {
				Assert.NotSame(hub1.ScopedDependency, hub2.ScopedDependency);
				Assert.Same(hub1.SingletonDependency, hub2.SingletonDependency);
			}

			using (var hub1 = (GenericLifetimeHubStub)hubManager.ResolveHub(nameof(GenericLifetimeHubStub)))
			using (var hub2 = (GenericLifetimeHubStub)hubManager.ResolveHub(nameof(GenericLifetimeHubStub))) {
				Assert.NotSame(hub1.ScopedDependency, hub2.ScopedDependency);
				Assert.Same(hub1.SingletonDependency, hub2.SingletonDependency);
			}
		}


		[Fact]
		public void DynamicHubOwnedDependenciesDisposeCorrectly()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			int scopedDisposalCount = 0;
			int singletonDisposalCount = 0;
			using (var hub = (DynamicLifetimeHubStub)hubManager.ResolveHub(nameof(DynamicLifetimeHubStub))) {
				hub.ScopedDependency.OnDisposing += (s, a) => scopedDisposalCount++;
				hub.SingletonDependency.OnDisposing += (s, a) => singletonDisposalCount++;
			}

			//The singleton dependency doesn't belong to the hub's scope
			Assert.Equal(0, singletonDisposalCount);

			//This dependency does and will be disposed when the hub is disposed
			Assert.Equal(1, scopedDisposalCount);
		}


		[Fact]
		public void GenericHubOwnedDependenciesDisposeCorrectly()
		{
			var container = SetupAutofacContainer();
			var hubManager = SetupSignalRHubManager(container);

			int scopedDisposalCount = 0;
			int singletonDisposalCount = 0;
			using (var hub = (GenericLifetimeHubStub)hubManager.ResolveHub(nameof(GenericLifetimeHubStub))) {
				hub.ScopedDependency.OnDisposing += (s, a) => scopedDisposalCount++;
				hub.SingletonDependency.OnDisposing += (s, a) => singletonDisposalCount++;
			}

			//The singleton dependency doesn't belong to the hub's scope
			Assert.Equal(0, singletonDisposalCount);

			//This dependency does and will be disposed when the hub is disposed
			Assert.Equal(1, scopedDisposalCount);
		}


		[Fact]
		public void UnmanagedLifetimeHubsDisposeWithoutThrowing()
		{
			new DynamicLifetimeHubStub(null, null).Dispose();
			new GenericLifetimeHubStub(null, null).Dispose();
		}

	}
}

File: Tests/SignalR.Extras.Autofac.Test/LifetimeHubFacts.fs
namespace SignalR.Extras.Autofac.Test

module LifetimeHubFacts =
    open System
    open System.Linq
    open SignalR.Extras.Autofac
    open SignalR.Extras.Autofac.Test.Stubs
    open Xunit


    [<Fact>]
    let ``SignalR Resolves Valid Lifetime Hubs``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        use hub = hubManager.ResolveHub("DynamicLifetimeHubStub")
        Assert.IsType<DynamicLifetimeHubStub>(hub) |> ignore
        Assert.True(hub :? ILifetimeHub) |> ignore
        Assert.NotNull(hub)

        use hub = hubManager.ResolveHub("GenericLifetimeHubStub")
        Assert.IsType<GenericLifetimeHubStub>(hub) |> ignore
        Assert.True(hub :? ILifetimeHub) |> ignore
        Assert.NotNull(hub)


    [<Fact>]
    let ``SignalR Resolves Valid Ordinary Hubs``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        use hub = hubManager.ResolveHub("DynamicOrdinaryHubStub")
        Assert.IsType<DynamicOrdinaryHubStub>(hub) |> ignore
        Assert.False(hub :? ILifetimeHub);
        Assert.NotNull(hub)

        use hub = hubManager.ResolveHub("GenericOrdinaryHubStub")
        Assert.IsType<GenericOrdinaryHubStub>(hub) |> ignore
        Assert.False(hub :? ILifetimeHub)
        Assert.NotNull(hub)


    [<Fact>]
    let ``SignalR Resolves Correct Number of Registered Hubs``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        let hubs = hubManager.ResolveHubs()
        Assert.Equal(4, hubs.Count())


    [<Fact>]
    let ``SignalR Resolves New Hub Instance Each Time``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        use hub1 = hubManager.ResolveHub("DynamicLifetimeHubStub")
        use hub2 = hubManager.ResolveHub("DynamicLifetimeHubStub")
        Assert.NotSame(hub1, hub2)      

        use hub1 = hubManager.ResolveHub("GenericLifetimeHubStub")
        use hub2 = hubManager.ResolveHub("GenericLifetimeHubStub")
        Assert.NotSame(hub1, hub2);


    [<Fact>]
    let ``Receives New Instances of Scoped Dependencies``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        use hub1 = hubManager.ResolveHub("DynamicLifetimeHubStub") :?> DynamicLifetimeHubStub
        use hub2 = hubManager.ResolveHub("DynamicLifetimeHubStub") :?> DynamicLifetimeHubStub
        Assert.NotSame(hub1.ScopedDependency, hub2.ScopedDependency)
        Assert.Same(hub1.SingletonDependency, hub2.SingletonDependency)

        use hub1 = hubManager.ResolveHub("GenericLifetimeHubStub") :?> GenericLifetimeHubStub
        use hub2 = hubManager.ResolveHub("GenericLifetimeHubStub") :?> GenericLifetimeHubStub
        Assert.NotSame(hub1.ScopedDependency, hub2.ScopedDependency)
        Assert.Same(hub1.SingletonDependency, hub2.SingletonDependency)


    [<Fact>]
    let ``Dynamic Hub Owned Dependencies Dispose Correctly``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        let scopedDisposalCount = ref 0
        let singletonDisposalCount = ref 0

        (
            use hub = hubManager.ResolveHub("DynamicLifetimeHubStub") :?> DynamicLifetimeHubStub
            hub.ScopedDependency.OnDisposing.Add(fun() -> scopedDisposalCount := !scopedDisposalCount + 1)
            hub.SingletonDependency.OnDisposing.Add(fun() -> singletonDisposalCount := !singletonDisposalCount + 1)
        )

        //The singleton dependency doesn't belong to the hub's scope
        Assert.Equal(0, !singletonDisposalCount)

        //This dependency does and will be disposed when the hub is disposed
        Assert.Equal(1, !scopedDisposalCount)


    [<Fact>]
    let ``Generic Hub Owned Dependencies Dispose Correctly``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        let scopedDisposalCount = ref 0
        let singletonDisposalCount = ref 0

        (
            use hub = hubManager.ResolveHub("GenericLifetimeHubStub") :?> GenericLifetimeHubStub
            hub.ScopedDependency.OnDisposing.Add(fun() -> scopedDisposalCount := !scopedDisposalCount + 1)
            hub.SingletonDependency.OnDisposing.Add(fun() -> singletonDisposalCount := !singletonDisposalCount + 1)
        )

        //The singleton dependency doesn't belong to the hub's scope
        Assert.Equal(0, !singletonDisposalCount)

        //This dependency does and will be disposed when the hub is disposed
        Assert.Equal(1, !scopedDisposalCount)


    [<Fact>]
    let ``Unmanaged Lifetime Hubs Dispose Without Throwing``() =
        let container = TestHelper.SetupAutofacContainer()
        let hubManager = TestHelper.SetupSignalRHubManager(container)

        use dynamicHub = hubManager.ResolveHub("DynamicLifetimeHubStub") :?> DynamicLifetimeHubStub
        use genericHub = hubManager.ResolveHub("GenericLifetimeHubStub") :?> GenericLifetimeHubStub

        dynamicHub.Dispose()
        genericHub.Dispose()
